src/canisters/frontend/ic-certified-assets/src/certification/types/certification.rs

```diff
@@ -82,24 +133,69 @@ impl HashTreePath {
             .map(|key| match key {
                 NestedTreeKey::String(k) => k.clone(),
                 NestedTreeKey::Bytes(b) => hex::encode(b),
+                NestedTreeKey::Hash(h) => hex::encode(h),
             })
             .collect::<Vec<String>>();
         let cbor = serialize_cbor_self_describing(&strings);
         base64::encode(cbor)
     }
+
+    pub fn new(
+        asset: &str,
+        status_code: u16,
+        headers: &[(String, Value)],
+        body: &RcBytes,
+        body_hash: Option<[u8; 32]>,
+    ) -> Self {
+        Self::from_parts(asset, status_code, headers, body, body_hash, None, None)
+    }
+
+    pub fn from_parts(
+        path: &str,
+        status_code: u16,
+        headers: &[(String, Value)],
+        body: &[u8],
+        body_hash: Option<[u8; 32]>,
+        certificate_expression: Option<&CertificateExpression>,
+        response_hash: Option<ResponseHash>,
+    ) -> Self {
+        let certificate_expression = certificate_expression
+            .cloned()
+            .unwrap_or_else(|| build_ic_certificate_expression_from_headers(headers));
+        let request_hash = RequestHash::default(); // request certification currently not supported
+        let body_hash = body_hash.unwrap_or_else(|| sha2::Sha256::digest(body).into());
+        let response_hash = response_hash
+            .unwrap_or_else(|| super::http::response_hash(headers, status_code, &body_hash));
+
+        AssetPath::from(path).hash_tree_path(&certificate_expression, &request_hash, response_hash)
+    }
+
+    pub fn not_found_base_path_v2() -> Self {
+        HashTreePath::from(Vec::from([
+            NestedTreeKey::String("http_expr".into()),
+            NestedTreeKey::String("<*>".into()),
+        ]))
+    }
+
+    pub fn not_found_base_path_v1() -> Self {
+        let not_found_path = AssetPath::from(FALLBACK_FILE);
+        not_found_path.asset_hash_path_v1()
+    }
 }
 
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub enum NestedTreeKey {
     String(String),
     Bytes(Vec<u8>),
+    Hash([u8; 32]),
```

Comment:
Maybe there are complicating factors here, but in general I prefer if `refactor:` PRs do not introduce new functionality or change existing functionality, but rather only restructure existing code.

Reply:
I agree, but in this case I would still consider this to be an internal restructuring of the code that is required for efficiency reasons (avoiding recomputing a response hash). That it still does the same as before is covered by unit tests.

If you want to call it an addition to the public interface I'm also totally ok not calling it a refactor anymore